crates/jstz_node/src/services/operations.rs

```
@@ -6,44 +12,120 @@ use axum::{
     Json,
 };
 
+use jstz_core::reveal_data::{PreimageHash, RevealData, MAX_REVEAL_SIZE};
 use jstz_core::BinEncodable;
-use jstz_proto::operation::{Operation, SignedOperation};
+use jstz_proto::operation::{Content, Operation, SignedOperation};
 use jstz_proto::receipt::Receipt;
+use octez::OctezRollupClient;
 use tezos_data_encoding::enc::BinWriter;
 use tezos_smart_rollup::inbox::ExternalMessageFrame;
 
 use utoipa_axum::router::OpenApiRouter;
 use utoipa_axum::routes;
 
+/// The maximum operation size in bytes that can be directly included without using the reveal mechanism.
+const MAX_DIRECT_OPERATION_SIZE: usize = 3915;
+
 pub struct OperationsService;
 
 const OPERATIONS_TAG: &str = "Operations";
 
 type HexEncodedOperationHash = String;
 
+// Given a large operation, encode it into preimages and store them in the rollup's preimages directory
+async fn prepare_rlp_operation(
+    rollup_client: &OctezRollupClient,
+    signer: KeyPair,
+    operation: &SignedOperation,
+    rollup_preimages_dir: &path::Path,
+) -> ServiceResult<SignedOperation> {
+    let reveal_type = operation
+        .verify_ref()
+        .map_err(|_| anyhow!("Invalid operation"))?
+        .content()
+        .try_into()
+        .map_err(|_| {
+            ServiceError::BadRequest("Large payload operation not supported".to_string())
+        })?;
+    let save_preimages = |hash: PreimageHash, preimage: Vec<u8>| {
+        let path = rollup_preimages_dir.join(hash.to_string());
+        fs::write(&path, preimage)
+            .unwrap_or_else(|_| panic!("failed to save preimage at: {:?}", path));
+    };
+    let KeyPair(public_key, secret_key) = signer;
+    let root_hash =
+        RevealData::encode_and_prepare_preimages(operation, save_preimages)
+            .map_err(|_| anyhow!("failed to prepare reval large payload operation"))?;
```

Not sure how this will affect the response message, but what about using `.context` here? A bit concerned about swallowing errors.